### ISSUE TITLE
Consider named axes via Haliax for neural network components

### ISSUE THREAD

**Issue by i-m-mll** [mlprt/feedbax#40]
## Context

Feedbax currently uses jaxtyping for shape/dtype annotations, with disciplined naming conventions for dimensions. This works but can become error-prone as neural network architectures grow in complexity—particularly with attention mechanisms, multi-head structures, or batch/sequence/feature interactions.

Named tensor libraries address this by making axis semantics explicit in the type system rather than relying on positional indices.

### Haliax (recommended if adopting named axes)

[Haliax](https://github.com/stanford-crfm/haliax) uses explicit `Axis` objects that carry both name and size:

```python
Head = hax.Axis("head", 8)
Key = hax.Axis("key", 64)
Pos = hax.Axis("position", 1024)

# Self-documenting operations
scores = hax.dot(query, key, axis=Key) / jnp.sqrt(Key.size)
```

**Strengths**:
- Axis objects carry size, enabling static validation
- Full FSDP and tensor parallelism support (proven to 70B parameters on TPU v4-2048)
- Works alongside Equinox modules
- Powers Stanford's Levanter/Marin for large-scale training

**Trade-offs**:
- Another abstraction layer over JAX arrays
- Requires axis-aware API calls (`hax.dot` instead of `jnp.dot`)
- Most valuable at scale; may be overkill for smaller models

### Penzai named axes (`pz.nx`)

Penzai offers a lighter-weight alternative using string names and `nmap` to lift functions:

```python
# Tag axes with names
named_arr = arr.tag("batch", "features")

# Lift any JAX function to vectorize over named axes
result = pz.nx.nmap(jnp.sum)(named_arr.untag("features"))
```

**Strengths**:
- Minimal API surface; works with existing JAX functions
- No redundant size storage (sizes live in the array, names are metadata)
- Easy to adopt incrementally

**Trade-offs**:
- Less explicit than Haliax (no static size in axis definition)
- Limited sharding/parallelism support
- Designed for interpretability workflows, not training at scale

## When Named Axes Help

1. **Complex broadcasting** - When positional semantics become ambiguous
2. **Attention mechanisms** - Query/key/value dimension tracking
3. **Multi-head structures** - Head dimension vs. feature dimension
4. **Sharding declarations** - Named axes map naturally to mesh axes
5. **Code reviews** - Self-documenting tensor operations

## When Named Axes Are Overkill

1. **Simple MLPs** - Dimension semantics are obvious
2. **Physics/mechanics code** - Physical quantities (angles, forces) are already semantic
3. **Tight dependency budget** - Another library to maintain
4. **Not hitting dimension bugs** - If current conventions work, don't fix what isn't broken

## Fit with Feedbax

Named axes would primarily benefit `SimpleStagedNetwork` and similar neural network components. The mechanics/physics code (`AbstractPlant`, `AbstractSkeleton`, `AbstractMuscle`) already has explicit physical semantics—joint angles, muscle lengths, forces—that don't benefit from named axes.

**Hybrid approach**: Use Haliax for NN internals while keeping standard JAX arrays for physics. This is viable since Haliax arrays can be converted to/from standard arrays at component boundaries.

## Recommendation

**Wait unless hitting dimension pain.** Current jaxtyping + naming conventions are adequate for Feedbax's scale. Revisit if:
- Attention mechanisms or transformers are added
- Dimension-related bugs become frequent
- Distributed training across multiple devices is needed

If adopting, **prefer Haliax** over Penzai's `pz.nx` due to:
- Explicit axis sizing
- Better sharding support
- Active development via Levanter/Marin ecosystem

## References

- [Haliax Documentation](https://haliax.readthedocs.io/en/latest/)
- [Haliax GitHub](https://github.com/stanford-crfm/haliax)
- [Levanter (uses Haliax + Equinox)](https://github.com/marin-community/levanter)
- [Penzai named axes](https://penzai.readthedocs.io/en/stable/notebooks/named_axes.html)
- ["Tensor Considered Harmful" (Alexander Rush)](https://nlp.seas.harvard.edu/NamedTensor)

---
Co-Authored-By: Claude Opus 4.5 <noreply@anthropic.com>
